### PR TITLE
Added Feature: Sync Command

### LINKED ISSUE #62: Add update meetings command
updates all meetings to update to next upcoming meeting

### PR DESCRIPTION
**Pull request by isanidiot**
Deletes meeting in the past, using the local time of machine as reference.
Closes #62

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/SyncCommand.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import static java.util.Objects.requireNonNull;
+import static seedu.address.model.Model.PREDICATE_SHOW_ALL_PERSONS;
+
+import seedu.address.model.Model;
+
+/**
+ * Syncs meetings to the local time on machine.
+ * Deletes meetings with times that are before the present time.
+ */
+public class SyncCommand extends Command {
```
Comment (ReaganTan00) [inline src/main/java/seedu/address/logic/commands/SyncCommand.java:12]: I think it should extend UndoableCommand, will be good to undo
Can implement in bugfix phase if needed

File: src/main/java/seedu/address/model/person/MeetingTimePastPredicate.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof MeetingTimePastPredicate // instanceof handles nulls
+                && now.equals(((MeetingTimePastPredicate) other).now)); // state check
+    }
+
+}
```
Comment (ReaganTan00) [inline src/main/java/seedu/address/model/person/MeetingTimePastPredicate.java:32]: 👍 Good implementation

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 3]
```diff
@@ -97,6 +97,13 @@ public void setPersons(List<Person> persons) {
         internalList.setAll(persons);
     }
 
+    /**
+     * Deletes meetings with times that are before the local time on machine.
+     */
+    public void syncMeetingTimes() {
+        internalList.forEach(Person::syncMeetingTimes);
+    }
+
```
Comment (ReaganTan00) [inline src/main/java/seedu/address/model/person/UniquePersonList.java:106]: Since this is done on the UniquePersonList, SyncCommand should work on the original list and not just the currently displayed list right?
Correct me if I'm wrong

File: src/test/java/seedu/address/logic/commands/SyncCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    private final Model model = new ModelManager(addressBook, new UserPrefs());
+    private final Model expectedModel = new ModelManager(addressBook, new UserPrefs());
+
+    @Test
+    void execute() {
+        assert(true);
+        SyncCommand command = new SyncCommand();
+        expectedModel.syncMeetingTimes();
+        assertCommandSuccess(command, model, SyncCommand.MESSAGE_SUCCESS, expectedModel);
+    }
```
Comment (ReaganTan00) [inline src/test/java/seedu/address/logic/commands/SyncCommandTest.java:30]: Good test case 👍

File: src/test/java/seedu/address/logic/parser/CreateCommandParserTest.java [hunk 5]
```diff
@@ -66,32 +66,32 @@ public void parse_allFieldsPresent_success() {
...
         assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_AMY
                 + ADDRESS_DESC_BOB + NETWORTH_DESC_BOB + MEETING_TIME_DESC_BOB + DESCRIPTION_DESC_BOB
-                + TAG_DESC_BOB, new CreateCommand(expectedPerson));
+                + TAG_DESC_AMY, new CreateCommand(expectedPerson));
 
         // multiple networths - last networth accepted
         assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB
                 + NETWORTH_DESC_AMY + NETWORTH_DESC_BOB + MEETING_TIME_DESC_BOB + DESCRIPTION_DESC_BOB
-                + TAG_DESC_BOB, new CreateCommand(expectedPerson));
+                + TAG_DESC_AMY, new CreateCommand(expectedPerson));
```
Comment (ReaganTan00) [inline src/test/java/seedu/address/logic/parser/CreateCommandParserTest.java:94]: Should be changed back to TAG_DESC_BOB, but doesn't affect the test cases. 
Likewise for the other tests

File: src/test/java/seedu/address/model/AddressBookTest.java [hunk 6]
```diff
@@ -83,6 +84,21 @@ public void getPersonList_modifyList_throwsUnsupportedOperationException() {
...
+        List<Person> newPersons = Arrays.asList(CARL, beta);
+        AddressBook newAddressBook = new AddressBook();
+        newAddressBook.setPersons(persons);
+        AddressBook expectedAddressBook = new AddressBook();
+        expectedAddressBook.setPersons(newPersons);
+        newAddressBook.syncMeetingTimes();
+
+        assertEquals(newAddressBook, expectedAddressBook);
+    }
+
```
Comment (ReaganTan00) [inline src/test/java/seedu/address/model/AddressBookTest.java:101]: Not sure if we have to test boundary values, that would be troublesome as would need to somehow create a stub LocalDateTime to represent the 'current' time.

File: src/test/java/seedu/address/model/person/MeetingTimePastPredicateTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        assertEquals(predicate, predicate);
+
+        // different types -> returns false
+        assertNotEquals(1, predicate);
+
+        // null -> returns false
+        assertNotEquals(null, predicate);
+
+    }
+}
```
Comment (ReaganTan00) [inline src/test/java/seedu/address/model/person/MeetingTimePastPredicateTest.java:69]: Good test cases to check if the predicate is working

File: src/test/java/seedu/address/model/person/PersonTest.java [hunk 8]
```diff
@@ -98,4 +99,29 @@ public void equals() {
...
+        Person deadGuy = new PersonBuilder().withMeetingTimes("28-07-2020-15:00", "15-06-2022-17:00").build();
+        Person.syncMeetingTimes(deadGuy);
+        assertEquals(deadGuy, new PersonBuilder().withMeetingTimes().build());
+
+        //Multiple meetings
+        Person busyGuy = new PersonBuilder().withMeetingTimes("14-03-2023-16:00", "28-07-2020-15:00",
+                "15-06-2022-17:00", "19-11-2024-02:00").build();
+        Person.syncMeetingTimes(busyGuy);
+        assertEquals(busyGuy, new PersonBuilder().withMeetingTimes("14-03-2023-16:00", "19-11-2024-02:00").build());
+    }
```
Comment (ReaganTan00) [inline src/test/java/seedu/address/model/person/PersonTest.java:126]: Good test cases that check past, present and future meeting times, as well as when multiple meeting times are checked.

**Review by ReaganTan00 — APPROVED** [review]
I am not sure if SyncCommand works on the original list or only current displayed list

Otherwise, LGTM with minor test fixes

File: src/main/java/seedu/address/model/person/UniquePersonList.java (hunk 3 above)
Comment (isanidiot) [reply src/main/java/seedu/address/model/person/UniquePersonList.java:106]: yep. it syncs the entire FABook

File: src/main/java/seedu/address/logic/commands/SyncCommand.java (hunk 1 above)
Comment (isanidiot) [reply src/main/java/seedu/address/logic/commands/SyncCommand.java:12]: I agree it would be good 💯 
But implementing it could be much more difficult than for other commands.
I suppose we can keep a copy of the uniquelist, tho that would probably break SLAP quite a bit
